### PR TITLE
Feed exercise implementation

### PR DESCRIPTION
**Pull request by ichachy**
First PR.
Would appreciate your notes.

### CODE REVIEW

File: app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt [hunk 1]
```diff
@@ -1,5 +1,24 @@
...
  */
+@JsonClass(generateAdapter = true)
+data class TemplatesMetadataItem(
+        @Json(name = "configuration") val configuration: String?,
+        @Json(name = "id") val id: String,
+        @Json(name = "isNew") val isNew: Boolean?,
+        @Json(name = "isPremium") val isPremium: Boolean,
+        @Json(name = "templateCategories") val templateCategories: List<String>?,
+        @Json(name = "templateName") val templateName: String?,
+        @Json(name = "templateThumbnailURI") val templateThumbnailURI: String)
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt]: Please add an empty line at the end of the file (eof).

File: app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt [hunk 2]
```diff
@@ -1,5 +1,24 @@
...
+data class GetFeedResponse(
+    @Json(name = "templatesMetadata") val templatesMetadataList: List<TemplatesMetadataItem>)
+
 /**
- * todo: add Data Transfer Object data class(es) here
+ * DTO class to hold the an instance of object in the JSON file
  */
+@JsonClass(generateAdapter = true)
+data class TemplatesMetadataItem(
+        @Json(name = "configuration") val configuration: String?,
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt]: This field is mandatory so it shouldn't be nullable. Same for the other fields of this class. So please remove the exclamation marks.

File: app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt [hunk 3]
```diff
@@ -1,5 +1,45 @@
...
+ * Holds a singleton FeedApiService object
  */
+interface FeedApiService{
+
+    @GET(SUFFIX_URL)
+    fun getFeed() : Single<GetFeedResponse>
+
+    companion object{
+        private const val BASE_URL = "https://assets.swishvideoapp.com/Android/demo/"
+        private const val SUFFIX_URL = "feed.json"
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt]: We need to found a better name for it. `FEED_URL`? We should also mark that this url is relative to the base url.

File: app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt [hunk 4]
```diff
@@ -1,5 +1,45 @@
...
+
+        /*
+         * Returns the single instance of the FeedApiService
+         */
+        fun getService(): FeedApiService{
+            if (feedApiService == null){
+                feedApiService = retrofit.create(FeedApiService::class.java)
+            }
+            return feedApiService!!
+        }
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt]: This method should be thread-safe. Since it might be accessed from any thread.

File: app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt [hunk 5]
```diff
@@ -1,5 +1,45 @@
...
+    }
+}
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt]: eof.

File: build.gradle [hunk 6]
```diff
@@ -1,11 +1,11 @@
 buildscript {
-    ext.kotlin_version = '1.3.72'
+    ext.kotlin_version = '1.4.31'
     repositories {
         google()
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.0.0'
+        classpath 'com.android.tools.build:gradle:4.1.1'
```
Comment (nissimLightricks) [inline build.gradle:8]: This is not related to implementing the network layer. So I would put it in a separate commit or even separate pr.

File: app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt [hunk 7]
```diff
@@ -1,5 +1,37 @@
...
+ * DAO class for FeedDatabase.
  */
+@Dao
+abstract class FeedDao{
+
+    @Query("SELECT * FROM feed_table")
+    abstract fun getAll(): Flowable<List<FeedEntity>>
+
+    @Query("SELECT COUNT(id) FROM feed_table")
+    abstract fun getTableSize(): Single<Int>
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt]: We should found a better name for this method. `getFeedEntitiesCount`?

File: app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt [hunk 8]
```diff
@@ -1,5 +1,37 @@
...
+        deleteAllSync()
+        insertSync(feedEntitiesList)
+    }
+
+    @Query("DELETE FROM feed_table")
+    protected abstract fun deleteAllSync()
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    protected abstract fun insertSync(feedEntitiesList: List<FeedEntity>)
+}
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt]: eof

File: app/src/main/java/com/lightricks/feedexercise/database/Entities.kt [hunk 9]
```diff
@@ -1,5 +1,16 @@
...
+import androidx.room.PrimaryKey
+
+
 /**
- * todo: add Room's Entity data class(es) here
+ * FeedItem Entity in the FeedDatabase
  */
+@Entity(tableName = "feed_table")
+data class FeedEntity(
+        @PrimaryKey val id: String,
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/Entities.kt]: Define explicitly the name for the id like you did for the other fields/columns to be consistent.

File: app/src/main/java/com/lightricks/feedexercise/database/Entities.kt [hunk 10]
```diff
@@ -1,5 +1,16 @@
...
+        @ColumnInfo(name = "thumbnailUrl") val thumbnailUrl: String,
+        @ColumnInfo(name = "isPremium") val isPremium: Boolean
+)
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/Entities.kt]: eof.

File: app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt [hunk 11]
```diff
@@ -1,5 +1,35 @@
...
+ * FeedDatabase class, expends RoomDatabase.
+ * Holds a singleton object of the database
  */
+@Database(entities = [FeedEntity::class], version = 1)
+abstract class FeedDatabase : RoomDatabase(){
+
+    abstract fun feedDao(): FeedDao
+
+    companion object{
+        private const val NAME = "feed_database"
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt]: We should found a better name. Name of what?

File: app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt [hunk 12]
```diff
@@ -1,5 +1,35 @@
...
+        fun getDatabase(context: Context): FeedDatabase{
+            synchronized(this){
+                if (database == null){
+                    database = Room.databaseBuilder(context.applicationContext,
+                        FeedDatabase::class.java, NAME)
+                        .build()
+                }
+                return database!!
+            }
+        }
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt]: Before acquiring the lock we should check for null. So in case it is non null we can return it directly without acquiring the lock at all.

Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/database/FeedDatabase.kt]: In Kotlin there is a built in solution for singleton object https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/lazy.html

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 13]
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
...
+    }
+
+    private val stateInternal: MutableLiveData<State> = MutableLiveData<State>(DEFAULT_STATE)
     private val networkErrorEvent = MutableLiveData<Event<String>>()
 
-    fun getIsLoading(): LiveData<Boolean> = isLoading
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
-    fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
+    private var disposable: Disposable? = null
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: Disposable of what? Please give it a more descriptive name.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 14]
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
...
+     * Represents the current state of the view model's data
+     */
+    data class State(
+        val feedItems: List<FeedItem>,
+        val isLoading: Boolean)
+
+    companion object{
+        private const val PREFIX_URL = "https://assets.swishvideoapp.com/" +
+                                            "Android/demo/catalog/thumbnails/"
+        private const val NETWORK_ERROR = "Network error has accrued"
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: typo: accrued -> occurred.

Comment (nissimLightricks) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: In general user facing strings are not hard coded but rather saved in a strings.xml and localised per languages.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 15]
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
...
+open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
+
+    /**
+     * Represents the current state of the view model's data
+     */
+    data class State(
+        val feedItems: List<FeedItem>,
+        val isLoading: Boolean)
+
+    companion object{
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: We usually place companion objects at the bottom of the class. Same for inner classes.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 16]
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
...
     init {
+        observeFeedRepository()
         refresh()
     }
 
+    /*
+     * Observes the FeedRepository and updates the stateInternal in case of any change
+     */
+    private fun observeFeedRepository() {
+        feedRepository.getAll().subscribeOn(Schedulers.io())
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: We should dispose the subscription once the user go away from the feature. This could be done by adding this reference to a `CompositeDisposable` and disposing it on `ViewModel#onCleared` method of the view model.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 17]
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
...
+    /**
+     * Refreshes the screen. Gets an updated data from the repository and updates
+     * the stateInternal accordingly
+     */
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        updateState { copy(isLoading = true) }
+        disposable?.dispose()
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: I would change the order of the two. By first disposing the previous task - so removing any effect it could has on the view model, then launching a new task. In this specific case there is no a real difference since the method and the callback are running on the main thread, however as a best practice I would change the order.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 18]
```diff
@@ -2,45 +2,108 @@ package com.lightricks.feedexercise.ui.feed
...
+        disposable = feedRepository.refresh().subscribeOn(Schedulers.io())
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: We save an instance of the disposable but we are never disposing it.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+import org.junit.runner.RunWith
+
+/**
+ * Tests for [FeedDaoTest] class.
+ */
+@RunWith(AndroidJUnit4::class)
+class FeedDaoTest {
+
+    @get:Rule
+    val instantExecutorRule = InstantTaskExecutorRule() // TODO ask Nissim
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: Please ask.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 20]
```diff
@@ -1,29 +1,74 @@
...
+    fun testRefreshAndThenGetAll_shouldReturnDataFromGetAll() {
+        val testObserver = feedRepository.refresh().andThen(feedRepository.getAll()).test()
+        testObserver.assertValue {
+            feedEntities ->
+            assertThat(feedEntities).isNotNull()
+            assertThat(feedEntities).isNotEmpty()
+            feedEntities == DummyFeedApiService.expectedFeedEntities
+        }
+    }
+}
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: eof

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        testObserver.dispose()
+    }
+
+    @Test
+    fun testGetAll_initialTableState_shouldNotReturnNull(){
+        val testObserver = feedDao.getAll().test()
+        testObserver.assertValue {
+            it != null
+        }
+    }
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: RxJava components never emit `null` so you can remove this test.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt (hunk 19 above)
Comment (nissimLightricks) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: I am not sure why we have this rule in the test. We can remove it.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    @get:Rule
+    val instantExecutorRule = InstantTaskExecutorRule() // TODO ask Nissim
+
+    private lateinit var feedDao: FeedDao
+
+    private lateinit var feedDatabase: FeedDatabase
+
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+
+    companion object{
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: Can you please move the companion object to the end of the class.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+        private val feedEntity1: FeedEntity = FeedEntity("1", "1", true)
+        private val feedEntity2: FeedEntity = FeedEntity("2", "2", true)
+        private val feedEntity3: FeedEntity = FeedEntity("3", "3", true)
+        private val feedEntity4: FeedEntity = FeedEntity("4", "4", true)
+
+        private val list1 = listOf(feedEntity1, feedEntity2)
+        private val list2 = listOf(feedEntity3, feedEntity4)
+        private val emptyList = emptyList<FeedEntity>()
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: What is the intent of this field?

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+
+    @After
+    fun closeDatabase(){
+        feedDatabase.close()
+    }
+
+    @Test
+    fun testGetTableSize_initialTableSize_shouldReturn0(){
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: Please remove the prefix `test` from all the method name. We already have this information that the method is a test with the `@Test` annotation.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+
+    @Test
+    fun testDeleteAllAndInsert_insertEmptyList_shouldChangeDatabaseAccordingly(){
+        val testObserver = feedDao.deleteAllAndInsert(emptyList).andThen(feedDao.getTableSize()).test()
+        testObserver.assertValue{
+                size -> size == 0
+        }
+        testObserver.dispose()
+    }
+}
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: eof

**Review by nissimLightricks — CHANGES_REQUESTED** [review]
Good job 👏. Round.

File: app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt (hunk 3 above)
Comment (ichachy) [reply app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt]: I changes the name to FEED_URL and added a comment that it's relative to base url

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt (hunk 16 above)
Comment (ichachy) [reply app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: Instead of using CompositeDisposable, I added another disposable variable. On onCleared I dispose both disposable variables of the ViewModel

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt (hunk 19 above)
Comment (ichachy) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: I Forgot to hide the TODO comment. Nir explained it to me last week.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt (hunk 23 above)
Comment (ichachy) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedDaoTest.kt]: I added a test to see  that the Dao object can handle an empty list in deleteAllAndInset method

**Review by nissimLightricks — APPROVED** [review]
Nice!
